### DISCUSSION TITLE
Audio Support with iMXRT1176

### DISCUSSION THREAD

**Discussion by aditya-jadoun1** [zephyrproject-rtos/zephyr#69332] [General]
I am using [RT1170EVK](https://www.nxp.com/design/design-center/development-boards/i-mx-evaluation-and-development-boards/i-mx-rt1170-evaluation-kit:MIMXRT1170-EVK) with Zephyr and trying to run I2S driver on it. I tried default sample application code (zephyr/samples/drivers/i2s/echo) and when building it is giving error bcoz it seems sample application is not configured to use with iMXRT1170/76.

_In function 'main':
Zephyr/zephyr/include/zephyr/device.h:85:41: error: '__device_dts_ord_DT_N_NODELABEL_i2s_rx_ORD' undeclared (first use in this function)
85 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                         ^~~~~~~~~

Zephyr/zephyr/include/zephyr/toolchain/common.h:137:26: note: in definition of macro '_DO_CONCAT'
  137 | #define _DO_CONCAT(x, y) x ## y
      |                          ^
Zephyr/zephyr/include/zephyr/device.h:85:33: note: in expansion of macro '_CONCAT'
   85 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                 ^~~~~~~
Zephyr/zephyr/include/zephyr/device.h:212:37: note: in expansion of macro 'DEVICE_NAME_GET'
  212 | #define DEVICE_DT_NAME_GET(node_id) DEVICE_NAME_GET(Z_DEVICE_DT_DEV_ID(node_id))
      |                                     ^~~~~~~~~~~~~~~
Zephyr/zephyr/include/zephyr/device.h:229:34: note: in expansion of macro 'DEVICE_DT_NAME_GET'
  229 | #define DEVICE_DT_GET(node_id) (&DEVICE_DT_NAME_GET(node_id))_

Is there any other sample code that I can directly run on the EVK to test audio or what changes I need to do in code, to make that example work for RT1170EVK?

**Comment by DerekSnell** [comment]
Hi @aditya-jadoun1 ,

NXP uses the [i2s_speed test](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/i2s/i2s_speed) to test the I2S driver on the MIMXRT1170-EVK.  That is a loopback test, so changes are needed to that board to connect the signals for the test to pass, see the [readme](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/i2s/i2s_speed).  

BTW, one trick I do to find samples and tests supported on specific boards is to search the repo for 'board name'.conf or 'board name'.overlay.  In this case, both [mimxrt1170_evk_cm7.conf](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/i2s/i2s_speed/boards/mimxrt1170_evk_cm7.conf) and [mimxrt1170_evk_cm7.overlay](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/i2s/i2s_speed/boards/mimxrt1170_evk_cm7.overlay) are included in the i2s_speed test.

Best regards

**Reply by aditya-jadoun1** [reply]
Hi @DerekSnell ,

Thanks for the reply. I tried that i2s_speed test example, but that example also getting failed. I am getting below error logs with that i2s_speed test example.
_

> Assertion failed at WEST_TOPDIR_Project/arm_mimxrt1170_evk_cm7_i2s_speed/src/test_i2s_speed.c:315: drivers_i2s_speed_test_i2s_transfer_long: ret not equal to 0
> 
> FAIL - test_i2s_transfer_long in 2.016 seconds
> 
> Assertion failed at WEST_TOPDIR_Project/arm_mimxrt1170_evk_cm7_i2s_speed/src/test_i2s_speed.c:211: drivers_i2s_speed_test_i2s_transfer_short: ret not equal to 0
> 
> FAIL - test_i2s_transfer_short in 0.015 seconds
> 
> START - test_i2s_dir_both_transfer_long
> I2S_DIR_BOTH value is not supported.
>  SKIP - test_i2s_dir_both_transfer_long in 0.004 seconds
> 
> SUITE FAIL -   0.00% [drivers_i2s_speed]: pass = 0, fail = 2, skip = 0, total = 2 duration = 2.031 seconds
>  - FAIL - [drivers_i2s_speed.test_i2s_transfer_long] duration = 2.016 seconds
>  - FAIL - [drivers_i2s_speed.test_i2s_transfer_short] duration = 0.015 seconds
> 
> PROJECT EXECUTION FAILED
> Failed after 1 attempts

_

Anyway, I tried some different sample code i2s_output found in the newer SDK and i am getting some output with that, but not the actual audio. I checked the I2S signals with NXP_MCU-SDK driver sample code (driver examples) and with Zephyr. I saw that MCLK is not getting generated properly with Zephyr driver. Attaching screenshots for the same and zephyr code i am working on. 

Please let me know if anything is wrong with that and why not getting SAI MCLK signal with zephyr i2s driver.

Thanks in advance.
[i2s_output_sample.zip](https://github.com/zephyrproject-rtos/zephyr/files/14533742/i2s_output_sample.zip)

[Saleae_Logs.zip](https://github.com/zephyrproject-rtos/zephyr/files/14533756/Saleae_Logs.zip)
![image](https://github.com/zephyrproject-rtos/zephyr/assets/160730720/469df91a-fb1e-4dde-ad97-1436bf912b9e)

![nxp_driver](https://github.com/zephyrproject-rtos/zephyr/assets/160730720/5c2135d6-19fd-4a41-96b0-102c6f257e6e)
![ZephyrI2sDriver](https://github.com/zephyrproject-rtos/zephyr/assets/160730720/319c6158-9459-47e1-a34e-b9dbda5c4813)

**Reply by RadekPolyend** [reply]
I was able to run tx or rx on this driver however I am not able to run configuration with both at the same time. Depending on what configuration you are using mclk may not be generated because the driver is not flexible enough i.e. it cannot generate mclk when it is in slave mode which is wrong because very often codec requires mclk to generate bclk and lrclk. There is a small change needed in the driver for this configuration to work and I think this driver should be modified in order to set mclk as output or input based on dts setting (dai/sai driver has it)

**Reply by DerekSnell** [reply]
> There is a small change needed in the driver for this configuration to work and I think this driver should be modified in order to set mclk as output or input based on dts setting (dai/sai driver has it)

Hi @RadekPolyend ,
Do you plan to contribute this change?  You could submit a PR to improve this driver.  Best regards

**Reply by RadekPolyend** [reply]
I can contribute such a change. Are there any specific prerequisites to be met that I can read about ?

**Reply by DerekSnell** [reply]
Thank you @RadekPolyend ,
Please review [Contributing to Zephyr](https://docs.zephyrproject.org/latest/contribute/index.html).  Thanks

**Comment by DerekSnell** [comment]
Hi @aditya-jadoun1 ,

I see you are right, and confirmed the i2s_speed test no longer passes on the MIMXRT1170_EVK.  I created an [Issue](https://github.com/zephyrproject-rtos/zephyr/issues/70051) about this, and found the last release when this test passes is Zephyr v3.3.0.  An update to the hal_nxp broke this test.  We will need to investigate further to find the root cause, and resolve this.

While that test does not pass, it does output the I2S signals.  The board [pinctrl for mimxrt1170_evk](https://github.com/zephyrproject-rtos/zephyr/blob/227462eda948a62776fba13a6c7e7470a6623f16/boards/nxp/mimxrt1170_evk/mimxrt1170_evk-pinctrl.dtsi#L228) configures the MCLK signal for SAI1.  Running the test, you can see MCLK on signal GPIO_AD_17, on J9 pin 15.

As a quick test, I also added the pinctrl setting below to the board overlay file to enable MCLK on the SAI4 pin.  This is signal GPIO_LPSR_05, connected to J10 pin 20.  So you can use these MCLK examples as a reference.
```
&pinmux_sai4 {
	group0 {
		pinmux = < &iomuxc_lpsr_gpio_lpsr_09_sai4_tx_data >,
		< &iomuxc_lpsr_gpio_lpsr_10_sai4_tx_sync >,
		< &iomuxc_lpsr_gpio_lpsr_12_sai4_tx_bclk >,
		< &iomuxc_lpsr_gpio_lpsr_05_sai4_mclk>;
	};
};
```
Best regards
